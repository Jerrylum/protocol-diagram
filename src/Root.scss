@use "style/Essential.scss";
@use "style/Modals.scss";

#root {
  overflow: hidden;

  --bg-body-color: #ffffff;
  --bg-default-color: #ffffff;
  --bg-paper-color: #f3f3f3;
  --text-primary-color: rgba(0, 0, 0, 0.87);
  --text-disabled-color: rgba(0, 0, 0, 0.6);
  --link-color: rgba(0, 0, 0, 0.87);
  --hover-color: rgba(0, 0, 0, 0.08);
  --focused-color: rgba(0, 0, 0, 0.12);
  background-color: var(--bg-body-color);

  > #root-container {
    display: flex;
    width: 100vw;
    width: 100svw;
    height: 100vh;
    height: 100svh;
    box-sizing: border-box;
  }
}
