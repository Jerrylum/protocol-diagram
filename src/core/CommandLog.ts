import { makeAutoObservable } from "mobx";

export type LogLevel = "info" | "error";

export interface CommandLog {
  level: LogLevel;
  message: string;
}

export class CommandLogger {
  private _logs: CommandLog[] = [];
  private _logCount = 0;

  constructor(public _maxLogs: number = 100) {
    makeAutoObservable(this);
  }

  get logs(): CommandLog[] {
    return this._logs;
  }

  get logCount(): number {
    return this._logCount;
  }

  add(level: LogLevel, message: string) {
    this._logs.push({ level, message });
    this._logs = this._logs.slice(-this._maxLogs);
    this._logCount++;
  }

  info(message: string) {
    this.add("info", message);
  }

  error(message: string) {
    this.add("error", message);
  }

  clear() {
    this._logs = [];
  }

  get maxLogs(): number {
    return this._maxLogs;
  }

  set maxLogs(maxLogs: number) {
    this._maxLogs = maxLogs;
    this._logs = this._logs.slice(-maxLogs);
  }
}
