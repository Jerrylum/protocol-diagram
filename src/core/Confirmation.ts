import { makeAutoObservable, when } from "mobx";
import { Modals } from "./Modals";

export interface ConfirmationButton {
  label: string;
  onClick?: () => void;
  hotkey?: string;
  color?: "inherit" | "primary" | "secondary" | "success" | "error" | "info" | "warning";
}

export interface ConfirmationPromptData {
  title: string;
  description: React.ReactNode;
  buttons: ConfirmationButton[];
  onKeyDown?(e: React.KeyboardEvent<HTMLDivElement>, onClick: (index: number) => void): void;
}

export interface ConfirmationInputPromptData extends ConfirmationPromptData {
  inputLabel: string;
  inputDefaultValue: string;
}

export class Confirmation {
  private data?: ConfirmationPromptData | ConfirmationInputPromptData;
  public input?: string;

  constructor(private modals: () => Modals) {
    makeAutoObservable(this);
  }

  close() {
    this.modals().close(ConfirmationModalSymbol);
    this.data = undefined;
  }

  async prompt(data: ConfirmationPromptData | ConfirmationInputPromptData, priority: number = 1) {
    if (data.buttons.length === 0) {
      data.buttons.push({ label: "OK" });
    }

    this.close();
    this.data = data;
    if ("inputLabel" in data && "inputDefaultValue" in data) {
      this.input = data.inputDefaultValue;
    } else {
      this.input = undefined;
    }

    this.modals().open(ConfirmationModalSymbol, priority);

    await when(() => this.data === undefined);

    return this.input;
  }

  get isOpen() {
    return this.data !== undefined && this.modals().opening === ConfirmationModalSymbol;
  }

  get title() {
    return this.data?.title ?? "";
  }

  set title(value: string) {
    if (this.data === undefined) return;
    this.data.title = value;
  }

  get description() {
    return this.data?.description ?? "";
  }

  set description(value: React.ReactNode) {
    if (this.data === undefined) return;
    this.data.description = value;
  }

  get buttons() {
    return this.data?.buttons ?? [];
  }

  get onKeyDown() {
    return this.data?.onKeyDown;
  }

  get inputLabel() {
    return (this.data as ConfirmationInputPromptData)?.inputLabel ?? "";
  }
}

export const ConfirmationModalSymbol = Symbol("ConfirmationModalSymbol");
