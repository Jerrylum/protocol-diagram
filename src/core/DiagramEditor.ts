import { makeAutoObservable } from "mobx";
import { clamp } from "./Util";
import { Vector } from "./Vector";

export class DiagramEditor {
  private _offset: Vector = new Vector(0, 0);
  private _scale: number = 1; // 1 = 100%, [1..3]

  get offset() {
    return this._offset;
  }

  get scale() {
    return this._scale;
  }

  set offset(offset: Vector) {
    this._offset = offset;
  }

  set scale(scale: number) {
    this._scale = clamp(scale, 0.75, 2);
  }

  constructor() {
    makeAutoObservable(this);
  }

  resetOffsetAndScale() {
    this.offset = new Vector(0, 0);
    this.scale = 1;
  }
}