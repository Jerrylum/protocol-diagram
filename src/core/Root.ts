import { Modals } from "./Modals";

export type RootStore = typeof rootStore;

const rootStore = {
  modals: new Modals()
} as const;

export function getRootStore(): RootStore {
  return rootStore;
}

