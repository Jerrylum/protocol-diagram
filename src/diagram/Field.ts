export class Field {
  static uidCount: number = 0;

  public name: string;
  public length: number;
  public uid: number;

  constructor(name: string, length: number, uid?: number) {
    this.name = name;
    this.length = length;
    this.uid = uid ?? Field.uidCount++;
  }

  equals(obj: unknown): boolean {
    // Especially for the uid comparison, ignore name and length
    if (obj instanceof Field) {
      return obj.uid === this.uid;
    }

    return false;
  }

  clone(): Field {
    return new Field(this.name, this.length, this.uid);
  }
}
