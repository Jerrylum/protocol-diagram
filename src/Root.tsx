import "./Root.scss";
import { observer } from "mobx-react-lite";
import { DiagramCanvas } from "./app/DiagramCanvas";
import { Box } from "@mui/material";
import { BottomPanel } from "./app/BottomPanel";
// import { TestModal, TestModalSymbol } from "./app/TestModal";
// import React from "react";
// import { getRootStore } from "./core/Root";

const Root = observer(() => {
  // React.useEffect(() => {
  //   getRootStore().modals.open(TestModalSymbol);
  // }, []);
  return (
    <Box id="root-container">
      <DiagramCanvas />
      <BottomPanel />
      {/* <TestModal /> */}
    </Box>
  );
});

export default Root;

