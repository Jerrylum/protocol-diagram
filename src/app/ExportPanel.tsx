import { Box, Button, Menu, MenuItem } from "@mui/material";
import { observer } from "mobx-react-lite";
import React from "react";
import { getRootStore } from "../core/Root";

export const ExportPanel = observer(() => {
  const { app } = getRootStore();

  const exportBtnRef = React.useRef<HTMLButtonElement>(null);
  const [isExportMenuOpen, setIsExportMenuOpen] = React.useState(false);

  const onExportMenuItemClick = () => {
    setIsExportMenuOpen(false);
  };

  const onExportAsText = () => {
    navigator.clipboard.writeText(app.diagram.toString());

    onExportMenuItemClick();
  };

  const onExportAsSVG = () => {
    navigator.clipboard.writeText(app.diagram.toSvgString());

    onExportMenuItemClick();
  };

  const onExportAsURL = () => {
    const encodedJsonDiagram = window.btoa(app.diagram.toJson());
    const base64String = encodedJsonDiagram.replaceAll("+", "-").replaceAll("/", "_");
    const origin = window.location.origin;
    const urlWithJson = `${origin}?diagram=${base64String}`;
    navigator.clipboard.writeText(urlWithJson);
    window.history.pushState({}, "", urlWithJson);
    onExportMenuItemClick();
  };

  return (
    <Box
      sx={{
        position: "absolute",
        top: "0",
        bottom: "0",
        left: "calc(100% + 4px)",
        width: "300px",
        display: "flex",
        alignItems: "center",
        gap: "4px"
      }}>
      <Button ref={exportBtnRef} onClick={() => setIsExportMenuOpen(true)}>
        Export
      </Button>
      <Menu
        anchorEl={exportBtnRef.current}
        open={isExportMenuOpen}
        onClose={onExportMenuItemClick}
        anchorOrigin={{ vertical: "top", horizontal: "center" }}
        transformOrigin={{ vertical: "bottom", horizontal: "center" }}>
        <MenuItem onClick={onExportAsText}>As Text</MenuItem>
        <MenuItem onClick={onExportAsSVG}>As SVG</MenuItem>
        <MenuItem onClick={onExportAsURL}>As URL</MenuItem>
      </Menu>
      {/* <Button>Share URL</Button> */}
    </Box>
  );
});
