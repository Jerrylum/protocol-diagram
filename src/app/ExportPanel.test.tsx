import { fireEvent, render } from "@testing-library/react";
import { ExportPanel } from "./ExportPanel";
import { act } from "react-dom/test-utils";
import { getRootStore } from "../core/Root";
import { Field } from "../diagram/Field";

test("Render ExportPanel", () => {
  // Create a mock clipboard object
  let mockClipboardData = "";
  const mockClipboard = {
    writeText: function (str: string) {
      mockClipboardData = str;
    },
    readText: function () {
      return mockClipboardData;
    }
  };
  // Define the new property
  Object.defineProperty(navigator, "clipboard", {
    get: () => mockClipboard
  });

  const components = (
    <div id="root-container">
      <ExportPanel />
    </div>
  );

  const result = render(components);
  expect(result.container).toMatchSnapshot();
  const button = result.container.querySelector("#root-container > div > button");

  // Test export as text
  act(() => {
    fireEvent.click(button!);
  });
  result.rerender(components);
  expect(result.baseElement).toMatchSnapshot();
  const exportTextli = result.baseElement.querySelectorAll("li")[0];
  const { app } = getRootStore();
  app.diagram.addField(new Field("a", 12));
  act(() => {
    fireEvent.click(exportTextli);
  });
  result.rerender(components);
  expect(navigator.clipboard.readText()).toBe(app.diagram.toString());

  // Test export as svg
  act(() => {
    fireEvent.click(button!);
  });
  result.rerender(components);
  expect(result.baseElement).toMatchSnapshot();
  const exportSvgli = result.baseElement.querySelectorAll("li")[1];
  act(() => {
    fireEvent.click(exportSvgli);
  });
  result.rerender(components);
  expect(navigator.clipboard.readText()).toBe(app.diagram.toSvgString());

  // Test export as url
  act(() => {
    fireEvent.click(button!);
  });
  result.rerender(components);
  expect(result.baseElement).toMatchSnapshot();
  const exportUrlli = result.baseElement.querySelectorAll("li")[2];
  act(() => {
    fireEvent.click(exportUrlli);
  });
  result.rerender(components);
  const encodedJsonDiagram = window.btoa(unescape(encodeURIComponent(app.diagram.toJson())));
  const base64String = encodedJsonDiagram.replaceAll("+", "-").replaceAll("/", "_");
  const origin = window.location.origin;
  const urlWithJson = `${origin}?diagram=${base64String}`;
  expect(navigator.clipboard.readText()).toBe(urlWithJson);
});
