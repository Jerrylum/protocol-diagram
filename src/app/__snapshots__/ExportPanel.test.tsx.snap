// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ExportPanel 1`] = `
<div>
  <div
    id="root-container"
  >
    <div
      class="MuiBox-root css-1v35qzk"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Export
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Render ExportPanel 2`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      id="root-container"
    >
      <div
        class="MuiBox-root css-1v35qzk"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Export
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiMenu-root MuiPopover-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiMenu-paper MuiPopover-paper MuiMenu-paper css-3dzjca-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          As Text
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          As SVG
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          As URL
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`Render ExportPanel 3`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      id="root-container"
    >
      <div
        class="MuiBox-root css-1v35qzk"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Export
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiMenu-root MuiPopover-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiMenu-paper MuiPopover-paper MuiMenu-paper css-3dzjca-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          As Text
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          As SVG
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          As URL
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`Render ExportPanel 4`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      id="root-container"
    >
      <div
        class="MuiBox-root css-1v35qzk"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Export
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiMenu-root MuiPopover-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiMenu-paper MuiPopover-paper MuiMenu-paper css-3dzjca-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          As Text
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          As SVG
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          As URL
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
